source/azure_iot_hub_client.c

```diff
@@ -626,3 +724,503 @@ AzureIoTHubClientResult_t AzureIoTHubClient_SendTelemetry( AzureIoTHubClient_t *
+    }
+    else
+    {
+        pxContext = &pxAzureIoTHubClient->_internal.xReceiveContext[ azureiothubRECEIVE_CONTEXT_INDEX_C2D ];
+        xMqttSubscription.xQoS = eAzureIoTMQTTQoS1;
+        xMqttSubscription.pcTopicFilter = ( const uint8_t * ) AZ_IOT_HUB_CLIENT_C2D_SUBSCRIBE_TOPIC;
+        xMqttSubscription.usTopicFilterLength = ( uint16_t ) sizeof( AZ_IOT_HUB_CLIENT_C2D_SUBSCRIBE_TOPIC ) - 1;
+        usSubscribePacketIdentifier = AzureIoTMQTT_GetPacketId( &( pxAzureIoTHubClient->_internal.xMQTTContext ) );
+
+        AZLogDebug( ( "Attemptting to subscribe to the MQTT topic: %s.", AZ_IOT_HUB_CLIENT_C2D_SUBSCRIBE_TOPIC ) );
+
+        if( ( xMQTTResult = AzureIoTMQTT_Subscribe( &( pxAzureIoTHubClient->_internal.xMQTTContext ),
+                                                    &xMqttSubscription, 1, usSubscribePacketIdentifier ) ) != eAzureIoTMQTTSuccess )
+        {
+            AZLogError( ( "Failed to subscribe, error=0x%08x", xMQTTResult ) );
+            xReturn = eAzureIoTHubClientFailed;
+        }
+        else
+        {
+            pxContext->_internal.usState = azureiothubTOPIC_SUBSCRIBE_STATE_SUB;
+            pxContext->_internal.usMqttSubPacketID = usSubscribePacketIdentifier;
+            pxContext->_internal.pxProcessFunction = prvAzureIoTHubClientC2DProcess;
+            pxContext->_internal.callbacks.xCloudToDeviceMessageCallback = xCallback;
+            pxContext->_internal.pvCallbackContext = prvCallbackContext;
+            xReturn = prvWaitForSubAck( pxAzureIoTHubClient, pxContext, ulTimeoutMilliseconds );
+        }
+    }
+
+    return xReturn;
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTHubClientResult_t AzureIoTHubClient_UnsubscribeCloudToDeviceMessage( AzureIoTHubClient_t * pxAzureIoTHubClient )
+{
+    AzureIoTMQTTSubscribeInfo_t xMqttSubscription = { 0 };
+    AzureIoTMQTTResult_t xMQTTResult;
+    AzureIoTHubClientResult_t xReturn;
+    uint16_t usSubscribePacketIdentifier;
+    AzureIoTHubClientReceiveContext_t * pxContext;
+
+    if( pxAzureIoTHubClient == NULL )
+    {
+        AZLogError( ( "Failed to unsubscribe cloud message: Invalid argument" ) );
+        xReturn = eAzureIoTHubClientInvalidArgument;
+    }
+    else
+    {
+        pxContext = &pxAzureIoTHubClient->_internal.xReceiveContext[ azureiothubRECEIVE_CONTEXT_INDEX_C2D ];
+        xMqttSubscription.xQoS = eAzureIoTMQTTQoS1;
+        xMqttSubscription.pcTopicFilter = ( const uint8_t * ) AZ_IOT_HUB_CLIENT_C2D_SUBSCRIBE_TOPIC;
+        xMqttSubscription.usTopicFilterLength = ( uint16_t ) sizeof( AZ_IOT_HUB_CLIENT_C2D_SUBSCRIBE_TOPIC ) - 1;
+        usSubscribePacketIdentifier = AzureIoTMQTT_GetPacketId( &( pxAzureIoTHubClient->_internal.xMQTTContext ) );
+
+        AZLogDebug( ( "Attemptting to unsubscribe from the MQTT topic: %s.", AZ_IOT_HUB_CLIENT_C2D_SUBSCRIBE_TOPIC ) );
+
+        if( ( xMQTTResult = AzureIoTMQTT_Unsubscribe( &( pxAzureIoTHubClient->_internal.xMQTTContext ),
```

Comment:
It appears this is a fire and forget, and we mark ourselves as unsubscribed before an UNSUBACK returns correct?

I like this simplicity, but are there scenarios that this will cause problems?  I'm not trying to make us add more code here assuming we don't need...

Suppose for instance app does SubscribeC2D / UnsubscribeC2D.  And suppose we send the bytes so this function succeeds, but server isn't happy for whatever reason.  And then app comes in to SubscribeC2D again.  Since they're already subscribed as far as server is concerned, would MQTT let the 2nd subscription to same topic just be "cool" or would there be an error?

Reply:
Yes, to keep the logic simple I am sending the unsubscribe and not waiting for it... Reason I am waiting for suback in subscribe path is: User may send QoS0 twin message and my never receive the response as topic was not sub ack before sending the GetTwin request.

Yup subscribing an existing subscribed topic is ok... I verfied it by sending multiple requests.

---
In reply to: [612745816](https://github.com/Azure/azure-iot-middleware-freertos/pull/75#discussion_r612745816) [](ancestors = 612745816)

Reply:
OK - the re-subscribe was only scenario that worried me here, so unless/until we can think of other issues agree with keeping this simple.